subsys/usb/class/bluetooth.c

```diff
@@ -251,6 +276,87 @@ static void bluetooth_status_cb(struct usb_cfg_data *cfg,
 	}
 }
 
+#define BT_HCI_ERR_VS_HANDLED  0xff
+
+#define CMD_EXT(_op, _min_len, _func) \
+	{ \
+		.op = _op, \
+		.min_len = _min_len, \
+		.func = _func, \
+	}
+
+static u8_t vs_read_usb_transport_mode(struct net_buf *buf)
+{
+	struct net_buf *rsp;
+	struct bt_hci_rp_vs_read_usb_transport_mode *rp;
+
+	rsp = bt_hci_cmd_complete_create(BT_HCI_OP_VS_READ_USB_TRANSPORT_MODE,
+				  sizeof(*rp) + 2);
+	rp = net_buf_add(rsp, sizeof(*rp));
+	rp->status = BT_HCI_ERR_SUCCESS;
+	rp->num_supported_modes = 2;
+
+	net_buf_add_u8(rsp, BT_HCI_VS_USB_H2_MODE);
+	net_buf_add_u8(rsp, BT_HCI_VS_USB_H4_MODE);
+
+	net_buf_put(&tx_queue, rsp);
+
+	return BT_HCI_ERR_VS_HANDLED;
+}
+
+static u8_t vs_set_usb_transport_mode(struct net_buf *buf)
+{
+	struct bt_hci_cp_vs_set_usb_transport_mode *cp;
+
+	cp = net_buf_pull_mem(buf, sizeof(*cp));
+
+	switch (cp->mode) {
+	case BT_HCI_VS_USB_H2_MODE:
+		mode = BT_RAW_MODE_NO_HEADER;
+		break;
+	case BT_HCI_VS_USB_H4_MODE:
+		mode = BT_RAW_MODE_H4;
+		break;
+	default:
+		LOG_DBG("Invalid mode: %u", cp->mode);
+		return BT_HCI_ERR_INVALID_PARAM;
+	}
+
+	LOG_DBG("mode %u", mode);
+
+	bt_raw_set_mode(mode);
+
+	return BT_HCI_ERR_SUCCESS;
+}
+
+static struct hci_cmd_ext {
+	u16_t  op;
+	size_t min_len;
+	u8_t   (*func)(struct net_buf *buf);
+} cmd_ext[] = {
+	CMD_EXT(BT_OCF(BT_HCI_OP_VS_READ_USB_TRANSPORT_MODE), 0,
+		vs_read_usb_transport_mode),
+	CMD_EXT(BT_OCF(BT_HCI_OP_VS_SET_USB_TRANSPORT_MODE),
+		sizeof(struct bt_hci_cp_vs_set_usb_transport_mode),
+		vs_set_usb_transport_mode),
+};
+
+static void vs_cmd_complete(u16_t op, u8_t status)
+{
+	struct net_buf *buf;
+	struct bt_hci_evt_cc_status *cc;
+
+	if (status == BT_HCI_ERR_VS_HANDLED) {
+		return;
+	}
+
+	buf = bt_hci_cmd_complete_create(op, sizeof(*cc));
+	cc = net_buf_add(buf, sizeof(*cc));
+	cc->status = status;
+
+	net_buf_put(&tx_queue, buf);
+}
+
 static int bluetooth_class_handler(struct usb_setup_packet *setup,
```

Comment:
I don't understand when this function is called. Will it still be called after switching to H4 mode? If not then it would seem that the vs_cmd handling will not be called anymore.

Reply:
This would be called by USB stack if host invokes [class request](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L1210)

Reply:
Right, if a vs command is sent over the bulk endpoint then this wouldn't so we have to handle both when receiving over the class request and when it sent over the bulk endpoint in case of H4 mode, so that needs fixing.